cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -324,6 +346,10 @@ class ARROW_EXPORT FileSystem : public std::enable_shared_from_this<FileSystem>
       const std::shared_ptr<const KeyValueMetadata>& metadata) = 0;
   Result<std::shared_ptr<io::OutputStream>> OpenAppendStream(const std::string& path);
 
+  using Factory = Result<std::shared_ptr<FileSystem>>(const Uri& uri,
```

Comment:
Style nit: we generally don't use nested type definitions except when they're used specifically in the enclosing class's APIs (which does not seem to be the case here). Should we make this a top-level `FileSystemFactory` instead?

Reply:
Alright